### PR TITLE
Hyperparameters

### PR DESCRIPTION
**Pull request by mizaqq**
-file tree changes
-moved most of the code out of notebooks
-refactoring of models

### CODE REVIEW

File: model/main.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+import argparse
+from model.utils.hyper_params import perform_random_search
+import logging
+import json
+
+logging.basicConfig(level=logging.INFO)
+logging.getLogger().setLevel(logging.INFO)
+
+argparser = argparse.ArgumentParser()
+argparser.add_argument("--model", type=str, default="regressor")
```
Comment (adamwawrzynski) [inline model/main.py]: I would put all code related to parsing CLI arguments into the function `def parse_args() -> argparse.Namespace:` for clarity.

File: model/main.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+
+def main(args):
+    df = pd.read_csv("model/data/train.csv")
```
Comment (adamwawrzynski) [inline model/main.py]: This hardcoded string should be changed to argparse argument with the default value.

File: model/main.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    df = preprocess_data(df)
+    # df = normalize_data(df)
+    X_train, X_test, y_train, y_test = split_data(df)
+    if args.model == "regressor":
+        model = models.Regressor()
+    elif args.model == "xgb":
+        with open('best_params.json', 'r') as f:
```
Comment (adamwawrzynski) [inline model/main.py]: Also here. All hardcoded strings/paths should be replaced with CLI arguments with the default value.

File: model/main.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+
+def main(args):
+    df = pd.read_csv("model/data/train.csv")
+    df = preprocess_data(df)
+    # df = normalize_data(df)
+    X_train, X_test, y_train, y_test = split_data(df)
+    if args.model == "regressor":
+        model = models.Regressor()
+    elif args.model == "xgb":
```
Comment (adamwawrzynski) [inline model/main.py]: I'm a fan of strict typing so I suggest changing the comparison between strings to enum class, such as:
```python
from enum import Enum

class ModelType(Enum):
    REGRESSOR: str =  "regressor"
    XGB: str = "xgb"
```
Then your if-else statement would look like this:
```python
model_type = ModelType(args.model)
if model_type.value == ModelType.REGRESSOR.value:
    ...
elif model_type.value == ModelType.XGB.value:
... 
```

File: model/main.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+            params = json.load(f)
+        model = models.XGBModel(**params)
+    elif args.model == "neural_network":
+        model = models.NeuralNetwork()
+        model.add_to_model(Dense(64, activation="relu"))
+        model.add_to_model(Dense(32, activation="relu"))
+        model.add_to_model(Dense(1, activation="linear"))
+        model.compile()
+    model.fit(X_train, y_train)
+    print(model.evaluate(X_test, y_test))
```
Comment (adamwawrzynski) [inline model/main.py]: When you use string comparison you are lacking the `else` statement, because you might pass the wrong string and it won't be matched. This will result in the `model` variable being not initialized. With the `Enum` class in such scenario, this line will throw an exception: `model_type = ModelType(args.model)`

File: model/main.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+
+args = argparser.parse_args()
+args.model = "xgb"
+main(args)
```
Comment (adamwawrzynski) [inline model/main.py]: I recommend not having code execution like this. You should add:
```python
if __name__ == "__main__":
   ...
```
By doing so you are enforcing code execution only if you run this script as an entry point for a Python interpreter. If you have code execution as you have now if you want to import this script at every `import` line this code will be executed. This is why it's a good practice not to have any code executed when importing (the exception will be singleton-like variable creation or something similar).

File: model/models/models.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+        self.model.fit(X, y, epochs=100, batch_size=32)
+
+    def predict(self, X: pd.DataFrame) -> pd.DataFrame:
+        return self.model.predict(X)
+
+    def evaluate(self, X, y: pd.DataFrame):
+        return r2_score(y, self.predict(X))
+
+
+model = XGBModel()
```
Comment (adamwawrzynski) [inline model/models/models.py]:
```suggestion
```

File: model/models/models.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+    def __init__(self, **kwargs):
+        super().__init__(model=LinearRegression())
+
+    def fit(self, X: pd.DataFrame, y: pd.DataFrame) -> None:
+        self.model.fit(X, y)
+
+    def predict(self, X: pd.DataFrame) -> pd.DataFrame:
+        return self.model.predict(X)
+
+    def evaluate(self, X, y: pd.DataFrame):
```
Comment (adamwawrzynski) [inline model/models/models.py]: Missing return type hint.
```suggestion
    def evaluate(self, X, y: pd.DataFrame) -> float:
```

File: model/models/models.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+    def fit(self, X: pd.DataFrame, y: pd.DataFrame):
+        pass
+
+    @abstractmethod
+    def predict(self, X: pd.DataFrame):
+        pass
+
+    @abstractmethod
+    def evaluate(self, predictions: pd.DataFrame, y: pd.DataFrame):
+        pass
```
Comment (adamwawrzynski) [inline model/models/models.py]: Missing return type hints.

File: model/models/models.py [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+        super().__init__(model=XGBRegressor())
+        self.model.set_params(**kwargs)
+
+    def fit(self, X: pd.DataFrame, y: pd.DataFrame) -> None:
+        self.model.fit(X, y)
+
+    def predict(self, X: pd.DataFrame) -> pd.DataFrame:
+        return self.model.predict(X)
+
+    def evaluate(self, X, y: pd.DataFrame):
```
Comment (adamwawrzynski) [inline model/models/models.py]: Missing return type hint.
```suggestion
    def evaluate(self, X, y: pd.DataFrame) -> float:
```

File: model/models/models.py [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+        return r2_score(y, self.predict(X))
+
+
+class NeuralNetwork(Model):
+    def __init__(self):
+        super().__init__(model=Sequential())
+
+    def add_to_model(self, layer: Layer):
```
Comment (adamwawrzynski) [inline model/models/models.py:73]: Missing return type hint.

File: model/models/models.py [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+        self.model.add(layer)
+
+    def compile(self, optimizer="adam", loss="mean_squared_error", metrics=["mean_squared_error"]):
```
Comment (adamwawrzynski) [inline model/models/models.py:76]: Missing return type hint.

File: model/models/models.py [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+        self.model.compile(optimizer=optimizer, loss=loss, metrics=metrics)
+
+# TODO refactor this to intake fit params
+    def fit(self, X: pd.DataFrame, y: pd.DataFrame) -> None:
+        self.model.fit(X, y, epochs=100, batch_size=32)
+
+    def predict(self, X: pd.DataFrame) -> pd.DataFrame:
+        return self.model.predict(X)
+
+    def evaluate(self, X, y: pd.DataFrame):
```
Comment (adamwawrzynski) [inline model/models/models.py]: Missing return type hint.

File: model/models/models.py [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+
+    def add_to_model(self, layer: Layer):
+        self.model.add(layer)
+
+    def compile(self, optimizer="adam", loss="mean_squared_error", metrics=["mean_squared_error"]):
+        self.model.compile(optimizer=optimizer, loss=loss, metrics=metrics)
+
+# TODO refactor this to intake fit params
+    def fit(self, X: pd.DataFrame, y: pd.DataFrame) -> None:
+        self.model.fit(X, y, epochs=100, batch_size=32)
```
Comment (adamwawrzynski) [inline model/models/models.py]: Use object fields to store and pass parameters instead of having hardcoded values. You can pass them in the constructor.

File: model/utils/data_analysis.py [hunk 15]
```diff
@@ -1,6 +1,6 @@
 import numpy as np
 import pandas as pd
-from utils.chi_square import perform_chi_square_test
+from model.utils.chi_square import perform_chi_square_test
```
Comment (adamwawrzynski) [inline model/utils/data_analysis.py]: Regarding the repository structure, it's a common practice to use `src` as a root directory for the code in the project.

File: model/utils/data_analysis.py [hunk 16]
```diff
@@ -12,7 +12,7 @@ def get_uncorrelated_col(df, threshold):
     for i in x.columns:
         corr_val = np.corrcoef(x[i], y)
         if abs(corr_val[0][1]) < threshold:
-            corr_y.append(i)
+            corr_y.append((i, round(float(corr_val[0][1]), 3)))
```
Comment (adamwawrzynski) [inline model/utils/data_analysis.py]: This might be converted into a one-liner (but not have to):
```python
corr_y = [(i, round(float(corr_val[0][1]), 3)) for corr_val in np.corrcoef(x[i], y)]
```

But what is more important you are rounding the correlation value to 3 digits. I think you should keep the full value in every step of your analysis for the accuracy of the results. You can round them when you want to print or visualize them.

File: model/utils/data_analysis.py [hunk 17]
```diff
@@ -22,7 +22,7 @@ def get_insignificant_columns(df):
     col_to_drop = []
     for sig in col_significance:
         if sig[1][2] == False:
-            col_to_drop.append(sig[0])
+            col_to_drop.append((sig[0], round(float(sig[1][1]), 3)))
```
Comment (adamwawrzynski) [inline model/utils/data_analysis.py]: Same here.

File: model/utils/data_preprocess.py [hunk 18]
```diff
@@ -72,3 +77,22 @@ def preprocess_data(df, df_test=None):
         df_test = encode_onehot(df_test)
         return df, df_test
     return df
+
+
+def scale_data(df):
```
Comment (adamwawrzynski) [inline model/utils/data_preprocess.py]: Missing type hints.

File: model/utils/data_preprocess.py [hunk 19]
```diff
@@ -72,3 +77,22 @@ def preprocess_data(df, df_test=None):
...
+    scaler = StandardScaler()
+    df = scaler.fit_transform(df)
+    return df
+
+
+def normalize_data(df):
```
Comment (adamwawrzynski) [inline model/utils/data_preprocess.py]: Missing type hints.

File: model/utils/data_preprocess.py [hunk 20]
```diff
@@ -72,3 +77,22 @@ def preprocess_data(df, df_test=None):
...
+    normalized_df = (df-df.mean())/df.std()
+    return normalized_df
+
+
+def split_data(df):
```
Comment (adamwawrzynski) [inline model/utils/data_preprocess.py]: Missing type hints.

File: model/utils/data_preprocess.py [hunk 21]
```diff
@@ -72,3 +77,22 @@ def preprocess_data(df, df_test=None):
...
+    X_train, X_test, y_train, y_test = train_test_split(
+        df.drop('SalePrice', axis=1),
+        df['SalePrice'],
+        test_size=0.2, random_state=0)
```
Comment (adamwawrzynski) [inline model/utils/data_preprocess.py]: I recommend using arguments with default values instead of hardcoded values.

File: model/utils/hyper_params.py [hunk 22]
```diff
@@ -0,0 +1,72 @@
...
+from sklearn.preprocessing import StandardScaler
+from sklearn.model_selection import train_test_split
+from sklearn.metrics import mean_squared_error, r2_score
+from xgboost import XGBRegressor
+from sklearn.model_selection import KFold, cross_val_score, RandomizedSearchCV, GridSearchCV
+import json
+
+
+def perform_grid_search():
+    df = pd.read_csv("./model/data/train_preprocessed.csv")
```
Comment (adamwawrzynski) [inline model/utils/hyper_params.py]: This should be a function argument with the default value.

File: model/utils/hyper_params.py [hunk 23]
```diff
@@ -0,0 +1,72 @@
...
+    scaler = StandardScaler()
+    X = scaler.fit_transform(df.drop("SalePrice", axis=1))
+    X_train, X_test, y_train, y_test = train_test_split(
+        df.drop("SalePrice", axis=1), df["SalePrice"], test_size=0.2, random_state=0
```
Comment (adamwawrzynski) [inline model/utils/hyper_params.py]: Same here.

File: model/utils/hyper_params.py [hunk 24]
```diff
@@ -0,0 +1,72 @@
...
+        "learning_rate": [0.01, 0.05, 0.1],
+        "lambda": [0.01, 0.1],
+        "alpha": [0.01, 0.1],
+    }
+
+    model_test = XGBRegressor(random_state=0)
+
+    skf = KFold(n_splits=5, shuffle=True, random_state=0)
+    grid = GridSearchCV(model_test, param_grid=params, scoring="r2", n_jobs=-1, cv=skf, verbose=4)
+    grid.fit(X_train, y_train)
```
Comment (adamwawrzynski) [inline model/utils/hyper_params.py]: Parameters should be a function argument with default values. Changing the function call parameter is easier than modifying the function's body every time you want to try a different configuration.

File: model/utils/hyper_params.py [hunk 25]
```diff
@@ -0,0 +1,72 @@
...
+    print("\n All results:")
+    print(grid.cv_results_)
+    print("\n Best estimator:")
+    print(grid.best_estimator_)
+    print("\n Best score:")
+    print(grid.best_score_ * 2 - 1)
+    print("\n Best parameters:")
+    print(grid.best_params_)
+    results = pd.DataFrame(grid.cv_results_)
+    results.to_csv("xgb-grid-search-results-01.csv", index=False)
```
Comment (adamwawrzynski) [inline model/utils/hyper_params.py]: Output path should definitely be the parameter of a function.

File: model/utils/hyper_params.py [hunk 26]
```diff
@@ -0,0 +1,72 @@
...
+
+    y_test = grid.best_estimator_.predict_proba(y_test)
+    results_df = pd.DataFrame(data={"id": y_test["id"], "target": y_test[:, 1]})
+    results_df.to_csv("submission-grid-search-xgb-porto-01.csv", index=False)
```
Comment (adamwawrzynski) [inline model/utils/hyper_params.py]: Same here.

File: model/utils/hyper_params.py [hunk 27]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    model_test = XGBRegressor(random_state=0)
+
+    skf = KFold(n_splits=5, shuffle=True, random_state=0)
+    param_comb = 7
+    random_search = RandomizedSearchCV(model_test, param_distributions=params,
+                                       n_iter=param_comb, scoring='r2', n_jobs=-1, cv=skf, verbose=3, random_state=0)
+    random_search.fit(X_train, y_train)
+    with open('best_params.json', 'w') as f:
```
Comment (adamwawrzynski) [inline model/utils/hyper_params.py]: Same here. Hardcoded values -> function parameters.

File: xgb-grid-search-results-01.csv
Comment (adamwawrzynski) [inline xgb-grid-search-results-01.csv:1]: Is it necessary in this PR?

**Review by adamwawrzynski — COMMENTED** [review]
Looks good. I left a couple of comments regarding missing type hints and hardcoded values that should be changed to the function parameters.

File: model/models/models.py [hunk 28]
```diff
@@ -68,15 +76,12 @@ def add_to_model(self, layer: Layer):
     def compile(self, optimizer="adam", loss="mean_squared_error", metrics=["mean_squared_error"]):
         self.model.compile(optimizer=optimizer, loss=loss, metrics=metrics)
 
-# TODO refactor this to intake fit params
-    def fit(self, X: pd.DataFrame, y: pd.DataFrame) -> None:
+    # TODO refactor this to intake fit params
+    def fit(self, X: pd.DataFrame, y: pd.DataFrame, epchos, batch_size) -> None:
         self.model.fit(X, y, epochs=100, batch_size=32)
```
Comment (adamwawrzynski) [inline model/models/models.py]:
```suggestion
    def fit(self, X: pd.DataFrame, y: pd.DataFrame, epchos: int = 100, batch_size: int = 32) -> None:
        self.model.fit(X, y, epochs=epochs batch_size=batch_size)
```

File: model/utils/data_preprocess.py [hunk 29]
```diff
@@ -61,7 +62,7 @@ def encode_onehot(df):
     return df
 
 
-def preprocess_data(df, df_test=None):
+def preprocess_data(df: pd.DataFrame, df_test=None) -> pd.DataFrame:
```
Comment (adamwawrzynski) [inline model/utils/data_preprocess.py]:
```suggestion
def preprocess_data(df: pd.DataFrame, df_test: Optional[pd.DataFrame] = None) -> pd.DataFrame:
```
I think it is `pd.DataFrame` type but I'm not sure.

**Review by adamwawrzynski — APPROVED** [review]
LGTM. Left comments regarding small improvements.

The title of PR is incorrect.
